celery_utils/models.py

```diff
@@ -60,3 +67,106 @@ def __str__(self):
+            self.completed_results.count(),
+            self.callback_result.status
+        )
+
+    def is_ready(self):
+        """
+        Check to see if all subtasks are in a finished state.
+        """
+        return all([result.status in READY_STATES for result in self.completed_results.all()])
+
+    def execute_callback(self):
+        """
+        Execute serialized callback. Called via on_chord_part_return.
+
+        There are no parameters, as everything we need is already serialized
+        in the model somewhere.
+        """
+        callback_signature = Signature.from_dict(json.loads(self.serialized_callback))
+
+        if any(result.status == FAILURE for result in self.completed_results.all()):
+            # we either remove the failures and only return results from successful subtasks, or fail the entire chord
+            if callback_signature.get('options', {}).get('propagate', current_app.conf.CELERY_CHORD_PROPAGATES):
+                try:
+                    raise ChordError(
+                        "Error in subtasks! Ids: {}".format([
+                            result.task_id
+                            for result in self.completed_results.all()
+                            if result.status == FAILURE
+                        ])
+                    )
+                except ChordError as error:
+                    self.mark_error(error, is_subtask=True)
+                    return
+            else:
+                """
+                Dev note: this doesn't *quite* match the behavior of the default backend.
+
+                According to http://www.pythondoc.com/celery-3.1.11/configuration.html#celery-chord-propagates,
+                the 2 options are to either propagate through (as done above), or to forward the Exception result
+                into callback (versus this approach of dropping error results). It seems silly to ask callbacks to
+                expect exception results as input though, so we drop them.
+                """  # pylint: disable=pointless-string-statement
+                for result in self.completed_results.all():
+                    if result.status == FAILURE:
+                        self.completed_results.remove(result)  # pylint: disable=no-member
+                        result.delete()
+
+        if callback_signature.get('options', {}).get('use_iterator', True):
+            # If we're using an iterator, it's assumed to be because there are size concerns with the results
+            # Thus, the callback_result TaskMeta object will have a null 'result' in the database, because you
+            # stored those results someplace else as part of the callback function, right?
+            def results_itr():
+                """Define results iterator."""
+                for subtask in self.completed_results.iterator():
+                    yield subtask.result
+            try:
+                callback_signature(results_itr)
```

Comment:
Couldn't you skip the `results_itr()` function, and just pass `callback_signature(self.completed_results.iterator)`?